pkg/kine/server/watch.go

```diff
@@ -4,49 +4,78 @@ import (
 	"context"
 	"sync"
 	"sync/atomic"
+	"time"
 
 	"github.com/sirupsen/logrus"
 	"go.etcd.io/etcd/api/v3/etcdserverpb"
 	"go.etcd.io/etcd/api/v3/mvccpb"
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"golang.org/x/exp/rand"
+	"k8s.io/apimachinery/pkg/util/wait"
 )
 
 var (
 	watchID int64
 )
 
+// explicit interface check
+var _ etcdserverpb.WatchServer = (*KVServerBridge)(nil)
+
+// getProgressReportInterval returns the configured progress report interval, with some jitter
+func (s *KVServerBridge) getProgressReportInterval() time.Duration {
+	// add rand(1/10*notifyInterval) as jitter so that kine will not
+	// send progress notifications to watchers at the same time even when watchers
+	// are created at the same time.
+	jitter := time.Duration(rand.Int63n(int64(s.limited.notifyInterval) / 10))
+	return s.limited.notifyInterval + jitter
+}
+
 func (s *KVServerBridge) Watch(ws etcdserverpb.Watch_WatchServer) error {
 	w := watcher{
-		server:  ws,
-		backend: s.limited.backend,
-		watches: map[int64]func(){},
+		server:   ws,
+		backend:  s.limited.backend,
+		watches:  map[int64]func(){},
+		progress: map[int64]chan<- int64{},
 	}
 	defer w.Close()
 
+	go wait.PollInfiniteWithContext(ws.Context(), s.getProgressReportInterval(), w.ProgressIfSynced)
+
 	for {
 		msg, err := ws.Recv()
 		if err != nil {
 			return err
 		}
 
-		if msg.GetCreateRequest() != nil {
-			w.Start(ws.Context(), msg.GetCreateRequest())
-		} else if msg.GetCancelRequest() != nil {
-			logrus.Debugf("WATCH CANCEL REQ id=%d", msg.GetCancelRequest().GetWatchId())
-			w.Cancel(msg.GetCancelRequest().WatchId, nil)
+		if cr := msg.GetCreateRequest(); cr != nil {
+			w.Start(ws.Context(), cr)
+		}
+		if cr := msg.GetCancelRequest(); cr != nil {
+			logrus.Tracef("WATCH CANCEL REQ id=%d", cr.WatchId)
+			w.Cancel(cr.WatchId, 0, 0, nil)
+		}
+		if pr := msg.GetProgressRequest(); pr != nil {
+			w.Progress(ws.Context())
 		}
 	}
 }
 
 type watcher struct {
-	sync.Mutex
+	sync.RWMutex
```

Comment:
Read-write mutexes have quite some overhead and are usually used only when there is a big number of readers and a very small number of writers. Is this there's only 2 cases where the lock is used:
 - to edit the watchers list (add/remove watcher)
 - to send a report.
 
 These two operations are mutual, so I think that this behaves just like a normal mutex, but with overhead.

Reply:
Looking at the upstream implementation there exists a RWlock on the progress report as well: https://github.com/etcd-io/etcd/blob/v3.5.11/server/mvcc/watchable_store.go#L54.

It is likely that we will have a high volume of progress reports, compared to small amount of watchers added/created for large deployments. In those cases the RWmutex would make sense to improve concurrency for readers.